server/perfkit/common/gae_cloud_sql_client.py

```diff
@@ -0,0 +1,155 @@
+  MySQLdb.FIELD_TYPE.TINY: 'INTEGER',
+  MySQLdb.FIELD_TYPE.SHORT: 'INTEGER',
+  MySQLdb.FIELD_TYPE.LONG: 'INTEGER',
+  MySQLdb.FIELD_TYPE.LONGLONG: 'INTEGER',
+  MySQLdb.FIELD_TYPE.FLOAT: 'FLOAT',
+  MySQLdb.FIELD_TYPE.DOUBLE: 'FLOAT',
+  MySQLdb.FIELD_TYPE.DATE: 'TIMESTAMP',
+  MySQLdb.FIELD_TYPE.DATETIME: 'TIMESTAMP',
+}
+
+
+class CloudSqlError(big_query_client.BigQueryError):
+  def __init__(self, message, query=None):
+    super(CloudSqlError, self).__init__(message, query)
+
+
+class GaeCloudSqlClient(object):
+  """Client for Cloud SQL from Appengine"""
+
+  shared_instance = None
+
+  def __init__(self, instance=None, db_name=None,
+               db_user=DB_USER, db_password=DB_PASSWORD):
+    if not instance:
+      raise CloudSqlError('Cloud SQL client: missing instance name')
+    if not db_name:
+      raise CloudSqlError('Cloud SQL client: missing db name')
+    if os.getenv('SERVER_SOFTWARE', '').startswith('Google App Engine/'):
+      self.db = MySQLdb.connect(
+          unix_socket='/cloudsql/' + instance, db=db_name, user=db_user,
+          passwd=db_password, charset='utf8')
+    else:
+      self.db = MySQLdb.connect(
+          host='127.0.0.1', port=3306, db=DB_NAME, user=DB_USER,
+          passwd=DB_PASSWORD, charset='utf8')
+    # Don't reuse the db connection - it doesn't seem to be thread
+    # safe. Getting results for the wrong query, and 'Lost
+    # Connection to MySQL server' errors. TODO(klausw): set up
+    # a per-backend connection pool.
+
+  # Here's an example of input data as returned by the Cloud SQL backend.
+  # We convert this to the same format used by the BigQuery backend.
+  #
+  # Input rows: (
+  #    (1456530028.0407979, u'GCP', u'n1-standard-64', u'us-central1-c', 8344.0),
+  #    (1456530088.4292591, u'GCP', u'n1-standard-64', u'us-central1-c', 8187.0),
+  #    (1456530039.241276, u'GCP', u'n1-standard-4', u'europe-west1-b', 5865.0))
+  # Tuples: (
+  #    ('timestamp', 5, 17, 22, 22, 31, 1),
+  #    ('cloud', 254, 3, 36, 36, 31, 1),
+  #    ('machine_type', 254, 14, 42, 42, 0, 1),
+  #    ('zone', 254, 14, 66, 66, 0, 1),
+  #    ('value', 5, 4, 22, 22, 31, 1))
+
+  def Query(self, query, timeout=None, cache_duration=None, use_cache=True):
+    cursor = self.db.cursor()
+    try:
+      cursor.execute(query)
+    except MySQLdb.Error as err:
+      message = err.message
+      # Extra paranoia since err.message isn't set consistently.
+      # Try to ensure we have a useful error getting reported.
+      if not message:
+        message = str(err)
+      logging.error('MySQL error: %s', message)
+      if not message:
+        # This shouldn't happen, but if for some reason we didn't
+        # get a message at all just re-throw and hope for the best.
+        # Converting to BigQueryError would lose stack information.
+        raise err
+      raise CloudSqlError(message, query)
+
+    rows_in = cursor.fetchall()
+    #logging.debug('Rows: %s', rows_in)
+    tuples = cursor.description
```

Comment:
Recommend factoring out the "transform to BQ format" to a separate method from "query Cloud SQL".